### PR TITLE
Add lancedb ner example

### PR DESCRIPTION
**Pull request by skrawcz**
See commits.

This is almost there expect for the example code -- need to also create a notebook for this to run properly.

## Changes
 - adds some materializers
 - adds union type check support 

## Checklist

- [ ] PR has an informative and human-readable title (this will be pulled into the release notes)
- [ ] Changes are limited to a single goal (no scope creep)
- [ ] Code passed the pre-commit check & code is left cleaner/nicer than when first encountered.
- [ ] Any _change_ in functionality is tested
- [ ] New functions are documented (with a description, list of inputs, and expected output)
- [ ] Placeholder code is flagged / future TODOs are captured in comments
- [ ] Project documentation has been updated if adding/changing functionality.

### CODE REVIEW

File: examples/LLM_Workflows/NER_Example/lancedb_module.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+
+def _write_to_lancedb(
+    data: Union[list[dict], pa.Table], db: lancedb.DBConnection, table_name: str
+) -> int:
+    """Helper function to write to lancedb.
+
+    This can handle the case the table exists or it doesn't.
+    """
+    try:
```
Comment (elijahbenizzy) [inline examples/LLM_Workflows/NER_Example/lancedb_module.py:23]: Is this the best way to do this with lance? No "does table exist"?

File: examples/LLM_Workflows/NER_Example/run.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+
+def load_data(table_name: str, use_tracker: bool = False):
+    adapter_list = [lifecycle.PrintLn()]
+    if use_tracker:
+        from hamilton_sdk import adapters
+
+        tracker = adapters.HamiltonTracker(
+            project_id=41,  # modify this as needed
+            username="elijah@dagworks.io",
```
Comment (elijahbenizzy) [inline examples/LLM_Workflows/NER_Example/run.py:28]: lol

File: examples/LLM_Workflows/NER_Example/run.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+from hamilton import driver, lifecycle
+
+
+def build_driver(adapter_list):
+    """Builds the driver with the necessary modules and adapters."""
+    dr = (
+        driver.Builder()
+        .with_config({})
+        .with_modules(ner_extraction, lancedb_module)
```
Comment (elijahbenizzy) [inline examples/LLM_Workflows/NER_Example/run.py:14]: I'm confused as to why these are lumped together. They share one node at the end but its confusing... We only call that node in one spot? 

Oh, it seems that they have shared resources (one way). 

IMO it is cleaner/easier to share if you break it up + it highlights the value of modularity + reuse for the blog post.

File: hamilton/plugins/huggingface_extensions.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+from hamilton.io import utils
+from hamilton.io.data_adapters import DataLoader, DataSaver
+
+COLUMN_FRIENDLY_DF_TYPE = False
+
+HF_types = (DatasetDict, Dataset, IterableDatasetDict, IterableDataset)
+
+
+@dataclasses.dataclass
+class HuggingFaceDSLoader(DataLoader):
```
Comment (elijahbenizzy) [inline hamilton/plugins/huggingface_extensions.py:38]: Blech all the insane huggingface params + code 😆

File: hamilton/plugins/huggingface_extensions.py [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+from hamilton import registry
+from hamilton.io import utils
+from hamilton.io.data_adapters import DataLoader, DataSaver
+
+COLUMN_FRIENDLY_DF_TYPE = False
+
+HF_types = (DatasetDict, Dataset, IterableDatasetDict, IterableDataset)
+
+
+@dataclasses.dataclass
```
Comment (elijahbenizzy) [inline hamilton/plugins/huggingface_extensions.py:37]: Comments on this -- why we have it, what it is (even though its obvious)

File: hamilton/plugins/huggingface_extensions.py [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+        else:
+            f_meta = (utils.get_file_metadata(self.path_or_buf),)
+        return {"file_metadata": f_meta, "dataset_metadata": ds_meta}
+
+    @classmethod
+    def name(cls) -> str:
+        return "parquet"
+
+
+if lancedb is not None:
```
Comment (elijahbenizzy) [inline hamilton/plugins/huggingface_extensions.py:142]: We should come up with a better way to do this... I guess this is fine for now, but it seems like we include this in lanceDB and only have the `applicable_types` include Dataset if we have Huggingface installed...

**Review by elijahbenizzy — COMMENTED** [review]
Looking good -- a few minor hygiene stuff

File: examples/LLM_Workflows/NER_Example/lancedb_module.py (hunk 1 above)
Comment (skrawcz) [reply examples/LLM_Workflows/NER_Example/lancedb_module.py:23]: yep.

File: examples/LLM_Workflows/NER_Example/run.py (hunk 3 above)
Comment (skrawcz) [reply examples/LLM_Workflows/NER_Example/run.py:14]: not sure I follow?

File: hamilton/plugins/huggingface_extensions.py (hunk 6 above)
Comment (skrawcz) [reply hamilton/plugins/huggingface_extensions.py:142]: that's the same though. So I thought this was more HF specific so should live here.

File: hamilton/plugins/huggingface_extensions.py [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+    streaming: bool = False
+    num_proc: Optional[int] = None
+    storage_options: Optional[Dict] = None
+    config_kwargs: Optional[Dict] = None
+
+    @classmethod
+    def applicable_types(cls) -> Collection[Type]:
+        return list(HF_types)
+
+    def _get_loading_kwargs(self) -> dict:
```
Comment (zilto) [inline hamilton/plugins/huggingface_extensions.py:67]: Might want to look at this approach: https://github.com/DAGWorks-Inc/hamilton/blob/818f2fca8b800b1a2b66fea31fdd73d5534beed8/hamilton/plugins/dlt_extensions.py#L96

I think `fields_to_skip` is a clear approach. Could document further why to skip in the docstring

File: examples/LLM_Workflows/NER_Example/run.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+        )
+        adapter_list.append(tracker)
+
+    dr = build_driver(adapter_list)
+
+    r = dr.execute(["lancedb_result"], inputs={"query": query, "table_name": table_name})
+    print(r)
+
+
+def main():
```
Comment (zilto) [inline examples/LLM_Workflows/NER_Example/run.py:64]: I believe `main()` should contain the logic and not rely on user inputs while `argparse` code exist after `if __name__ == "__main__"`.

File: examples/LLM_Workflows/NER_Example/run.py (hunk 3 above)
Comment (elijahbenizzy) [reply examples/LLM_Workflows/NER_Example/run.py:14]: Specifically there are two drivers. They do two different things. One has a dependency on the other. The graph is the same/there is one `build_driver` function.

Feels like we should break it into two?

File: hamilton/plugins/huggingface_extensions.py (hunk 7 above)
Comment (skrawcz) [reply hamilton/plugins/huggingface_extensions.py:67]: I tried it -- and it's more lines of code in the other cases. So not going to use it here.

**Review by elijahbenizzy — APPROVED** [review]
Looks fine, I think we should probably break it up into more fns/modules, but see it going either way

**Pull request merged**